### PR TITLE
[MDEP-708] - dependency:analyze recommends test scope for test-only artifacts that have non-test scope

### PR DESCRIPTION
**Pull request by ian-lavallee**
Works similarly to the current logic for getting unused dependencies and declared; looks in output src and test folders and compares what it finds.

Tried to make the test less complicated with only one pom but couldn't get it working.

### CODE REVIEW

**Comment by ian-lavallee** [comment]
@elharo

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 1]
```diff
@@ -129,6 +135,21 @@ public ProjectDependencyAnalysis analyze( MavenProject project )
         return results;
     }
 
+    protected Set<Artifact> getTestArtifactsWithNonTestScope( Set<Artifact> testOnlyArtifacts )
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:138]: can this be private or package protected?

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 2]
```diff
@@ -189,6 +210,29 @@ else if ( file != null && file.isDirectory() )
         return artifactClassMap;
     }
 
+    protected Set<String> buildTestDependencyClasses( MavenProject project ) throws IOException
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:213]: can this be private or package protected?

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 3]
```diff
@@ -189,6 +210,29 @@ else if ( file != null && file.isDirectory() )
...
+    {
+        Set<String> nonTestDependencyClasses = new HashSet<String>();
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:215]: Can just write `new HashSet<>` without an explicit generic type

File: src/main/java/org/apache/maven/shared/dependency/analyzer/ProjectDependencyAnalysis.java [hunk 4]
```diff
@@ -44,19 +44,23 @@
...
+
     // constructors -----------------------------------------------------------
 
     public ProjectDependencyAnalysis()
     {
-        this( null, null, null );
+        this( null, null, null, null );
     }
 
     public ProjectDependencyAnalysis( Set<Artifact> usedDeclaredArtifacts, Set<Artifact> usedUndeclaredArtifacts,
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/ProjectDependencyAnalysis.java:56]: adding an argument incompatibly changes the existing API; consider an additional constructor instead

File: src/main/java/org/apache/maven/shared/dependency/analyzer/ProjectDependencyAnalysis.java [hunk 5]
```diff
@@ -88,6 +92,15 @@ public ProjectDependencyAnalysis( Set<Artifact> usedDeclaredArtifacts, Set<Artif
         return unusedDeclaredArtifacts;
     }
 
+    /**
+     * Test Artifacts that have a non-test scope
+     * @return {@link Artifact}
+     */
+    public Set<Artifact> getTestArtifactsWithNonTestScope()
+    {
+        return testArtifactsWithNonTestScope;
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/ProjectDependencyAnalysis.java:101]: exposes mutable internal state; a defensive copy might be appropriate

File: src/test/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzerTest.java [hunk 6]
```diff
@@ -267,12 +271,36 @@ public void testJarWithXmlTransitiveDependency()
...
+    public void testJarWithNonTestScopedTestDependency()
+            throws TestToolsException, ProjectDependencyAnalyzerException
+    {
+        compileProject( "jarWithNonTestScopedTestDependency/pom.xml" );
+
+        MavenProject project2 = getProject( "jarWithNonTestScopedTestDependency/project2/pom.xml" );
+
+        ProjectDependencyAnalysis actualAnalysis = analyzer.analyze( project2 );
+
+        Artifact project1 = createArtifact( "org.apache.maven.shared.dependency-analyzer.tests",
```
Comment (elharo) [inline src/test/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzerTest.java:290]: artifact1?

File: src/test/resources/jarWithNonTestScopedTestDependency/pom.xml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+<project
+	xmlns="http://maven.apache.org/POM/4.0.0"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
+>
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.apache.maven.shared.dependency-analyzer.tests</groupId>
+	<artifactId>jarWithNonTestScopedTestDependency</artifactId>
+	<packaging>pom</packaging>
```
Comment (elharo) [inline src/test/resources/jarWithNonTestScopedTestDependency/pom.xml:30]: 2 space indents in XML

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java (hunk 2 above)
Comment (ian-lavallee) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:213]: All the other buildMethod methods are protected so I made this protected to keep it the same

Comment (elharo) [reply src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:213]: It is very common to make methods overly open. Thus I wouldn't assume that the existing methods should have been protected. We can always make private methods more public if we discover a need. It is much harder to make a public method private again.

File: src/test/resources/jarWithNonTestScopedTestDependency/pom.xml (hunk 7 above)
Comment (ian-lavallee) [reply src/test/resources/jarWithNonTestScopedTestDependency/pom.xml:30]: That's weird it is 2 spaces in my IDE. I'll look into it

Comment (elharo) [reply src/test/resources/jarWithNonTestScopedTestDependency/pom.xml:30]: then it's probably tabs, which you absolutely should not use

**Comment by ian-lavallee** [comment]
Thought we might run into this. The dependency identification fails on test dependencies when using Java 7. Last commit should fix the issue, I'll monitor the CI

**Comment by ian-lavallee** [comment]
Actually @elharo need you to trigger a build for me.

**Comment by elharo** [comment]
Jenkins passed.

**Pull request merged**